client/src/stellar_error.rs

```diff
@@ -0,0 +1,244 @@
+use std::error::Error;
+use serde::{de, Deserialize, Deserializer};
+use std::fmt;
+
+/// A resource for the stellar horizon API specific error codes.
+/// These errors adhere to the Problem Details Standard
+/// https://tools.ietf.org/html/draft-ietf-appsawg-http-problem-00
+/// and a list of possible erros can be found here
+/// https://www.stellar.org/developers/horizon/reference/errors.html
+#[derive(Debug)]
+pub struct StellarError {
+    // type is a protected word
+    error_type: ErrorType,
+    title: String,
+    status: u16,
+    detail: String,
+    instance: String,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ErrorType {
+    BadRequest,
+    BeforeHistory,
+    Forbidden,
+    NotAcceptable,
+    NotFound,
+    NotImplemented,
+    RateLimitExceeded,
+    InternalServerError,
+    StaleHistory,
+    TransactionFailed,
+    TransactionMalformed,
+    NotAValidErrorType,
+}
+
+#[derive(Deserialize)]
+struct IntermediateStellarError {
+    // type is a protected word
+    #[serde(rename = "type")] error_type: String,
+    title: String,
+    status: u16,
+    detail: String,
+    instance: String,
+}
+
+impl<'de> Deserialize<'de> for StellarError {
+    fn deserialize<D>(d: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let rep: IntermediateStellarError = IntermediateStellarError::deserialize(d)?;
+        StellarError::new(
+            &rep.error_type,
+            rep.title,
+            rep.status,
+            rep.detail,
+            rep.instance,
+        ).map_err(|err| de::Error::custom(err))
+    }
+}
+
+impl StellarError {
+    /// Creates a new stellar error.  If the error type is not recognized an error is returned.
+    pub fn new(
+        error_type: &str,
+        title: String,
+        status: u16,
+        detail: String,
+        instance: String,
+    ) -> Result<StellarError, String> {
+        let type_enum: ErrorType = match error_type {
+            "bad_request" => ErrorType::BadRequest,
+            "before_history" => ErrorType::BeforeHistory,
+            "forbidden" => ErrorType::Forbidden,
+            "not_acceptable" => ErrorType::NotAcceptable,
+            "not_found" => ErrorType::NotFound,
+            "not_implemented" => ErrorType::NotImplemented,
+            "rate_limit_exceeded" => ErrorType::RateLimitExceeded,
+            "internal_server_error" => ErrorType::InternalServerError,
+            "stale_history" => ErrorType::StaleHistory,
+            "transaction_failed" => ErrorType::TransactionFailed,
+            "transaction_malformed" => ErrorType::TransactionMalformed,
+            _ => ErrorType::NotAValidErrorType,
```

Comment:
I think accepting this as an `Other(String)` might be better so that we can keep the value if they add an additional type